### PR TITLE
(WIP) GMG: use deal.II compute_diagonal()

### PR DESCRIPTION
**Pull request by tjhei**
Use compute_diagonal() inside deal.II instead of our own version. This needs to be tested more, but we would be able to get rid of a lot of code this way.

I am opening this to be able to run the tests.

FYI: @zjiaqi2018

### CODE REVIEW

**Comment by peterrum** [comment]
Why are the test outputs changing? Is it due to different values on the constrained entries or is this just a numerical artifact?

**Comment by tjhei** [comment]
I am not sure, but this could be a bug. It seems like it only makes a difference if the Newton solver is used.

**Comment by tjhei** [comment]
@zjiaqi2018 Can you take a look and review this PR?

**Comment by zjiaqi2018** [comment]
> @zjiaqi2018 Can you take a look and review this PR?

Sure

**Comment by zjiaqi2018** [comment]
> I am not sure, but this could be a bug. It seems like it only makes a difference if the Newton solver is used.

Strange, ABlock doesn't even have Newton derivatives.

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -2928,139 +2904,14 @@ namespace aspect
...
-    const bool is_compressible = sim.material_model->is_compressible();
-
-    // Assemble and store the diagonal of the GMG level matrices derived from:
-    // 2*eta*(symgrad u, symgrad v) - (if compressible) 2*eta/3*(div u, div v)
     for (unsigned int level=0; level < sim.triangulation.n_global_levels(); ++level)
       {
         mg_matrices_Schur_complement[level].compute_diagonal();
 
-        // If we have a tangential boundary we must compute the A block
-        // diagonal outside of the matrix-free object
```
Comment (zjiaqi2018) [inline source/simulator/stokes_matrix_free.cc:2940]: Just curious, why do we need to do this before?

File: source/simulator/stokes_matrix_free.cc [hunk 2]
```diff
@@ -2928,139 +2904,14 @@ namespace aspect
...
-          }
-        else
-          {
-            mg_matrices_A_block[level].compute_diagonal();
-          }
+        mg_matrices_A_block[level].compute_diagonal();
 
         // This vector is no longer needed. Resize to 0.
-        level_viscosity_vector[level].reinit(0);
+        //level_viscosity_vector[level].reinit(0);
```
Comment (zjiaqi2018) [inline source/simulator/stokes_matrix_free.cc:2914]: I guess we can remove these two lines, and move `level_viscosity_vector[level].reinit(0);`  to `evaluate_material_model()` after `mg_matrices_A_block[level].set_cell_data (level_cell_data[level]);`

**Comment by tjhei** [comment]
superseded by #5060 #5061 #5062

**Pull request closed without merging**